### DISCUSSION TITLE
No output on "Process completed with exit code 1"

### DISCUSSION THREAD

**Discussion by Senui** [community/community#25939] [Actions]
<p>Hello,</p>
<p>One of my workflows is failing (see [1]), but I don’t seem to get any valuable output regarding the cause of the failure. Unfortunately, this is not the first time this has happened, and I had to spend hours trying to debug my changes.</p>
<p>I set both <code>ACTIONS_RUNNER_DEBUG</code> and <code>ACTIONS_STEP_DEBUG</code> to <code>true</code> in the hope that I could get some more insight, but nothing useful…</p>
<p>It seems like the runner takes my bash commands and generates a temporary shell script to execute them all. This however seems to sometimes hide the error message when one of the commands fails.</p>
<p>Is there any way to debug this, or at least get more output?</p>
<p>Currently, I get this as the output:</p>
<pre><code class="lang-auto">Run . scl_source enable devtoolset-7 || true
  . scl_source enable devtoolset-7 || true
  . /etc/profile.d/modules.sh || true
  module load mpi
  . $INSTALL_DIR/bin/thisbdm.sh
  git config --system user.name "Test User"
  git config --system user.email user@test.com
  xvfb-run -d -s "-screen 0 2560x1440x24" make run-demos
  shell: bash --noprofile --norc -e -o pipefail {0}
  env:
    INSTALL_DIR: /github/home/biodynamo-v0.1.0
##[debug]bash --noprofile --norc -e -o pipefail /__w/_temp/5b25a4b0-779d-4218-b1cd-80de8a7a01e8.sh
##[error]Process completed with exit code 1.
##[debug]Finishing: System tests BioDynaMo
</code></pre>
<p>[1] <a href="https://github.com/BioDynaMo/biodynamo/runs/893818544?check_suite_focus=true" rel="nofollow noopener">https://github.com/BioDynaMo/biodynamo/runs/893818544?check_suite_focus=true</a></p>
<p>Best,<br>
Ahmad</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/senui">@senui</a>,</p>
<p>I forked your repository and tested. According to my troubleshooting, I found the error returned from the command line “<strong>$INSTALL_DIR/bin/thisbdm.sh</strong>”.<br>
The error did not occur on this command line itself. It should be returned by a certain command line in the script ‘<strong>thisbdm.sh</strong>’.</p>
<p>I recommend that you can try to check if the command lines in the script ‘<strong>thisbdm.sh</strong>’ contain some defects, and also try executing this script on your local environment to see if it can work fine.</p>

**Comment by Senui** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>,</p>
<p>Thanks for your reply. Could you please tell me how you managed to find out that the error originated from <code>$INSTALL_DIR/bin/thisbdm.sh</code>?</p>
<p>We are running the same type workflow on Ubuntu 18.04, 20.04, and on macOS, but without any failures there…</p>
<p>Yes, I could check the command lines in the <code>thisbdm.sh</code> script one-by-one and I could try executing the script locally to debug this, but as I tried to make clear in my opening post, this could potentially cost me hours and hours of work.</p>
<p>Why is it not possible to see in the Github Actions output which line of <code>thisbdm.sh</code> returned a non-zero exit code? If you run any bash script normally in a terminal, it should output the error message, and not try to hide it.</p>
<p>I hope that there is some more convenient way to debug this vague failure.</p>
<p>Edit:<br>
I just tried running the exact same commands in the same CentOS docker container that the failing workflow is using, and I cannot reproduce the error :frowning:</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/senui">@senui</a>,</p>
<blockquote>
<p>Could you please tell me how you managed to find out that the error originated from  <code>$INSTALL_DIR/bin/thisbdm.sh</code> ?</p>
</blockquote>
<p>In the “System tests BioDynaMo” step, I printed a label after each command line executed successfully, and found the label for the line “<strong>$INSTALL_DIR/bin/thisbdm.sh</strong>” did not print. So I could locate the command line where the error was returned from.</p>
<p>For this issue, I have created an issue ticket (<a href="https://github.com/actions/virtual-environments/issues/1279" rel="nofollow noopener">actions/virtual-environments#1279</a>) to help you report it.<br>
You can follow this issue ticket and add you comment on it.</p>

**Comment by Senui** [comment]
<p>Thanks for opening the issue ticket. I hope the team understands that the absence of an error message makes this a non-trivial debugging issue. I will keep an eye on the ticket</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/senui">@senui</a>,</p>
<p>OK, please keep following the issue ticket.</p>

**Comment by al-cheb** [comment, marked as answer]
<p>Hello, <a class="mention" href="/senui">@senui</a></p>
<p>I am working on the ticket. I set <code>set -x</code> to find a line that cause the issue:</p>
<pre><code class="lang-auto"> - name: System tests BioDynaMo
       shell: bash
       working-directory: build
       run: |
         set -x
</code></pre>
<p>Line:</p>
<pre><code class="lang-auto">2020-07-23T11:30:35.2865130Z +++ /usr/bin/scl_enabled llvm-toolset-6.0
2020-07-23T11:30:35.3275453Z ##[error]Process completed with exit code 1.
</code></pre>
<p>I found out which part of code is executing this line:</p>
<ol>
<li><a href="https://github.com/BioDynaMo/biodynamo/blob/5092924f4caf7a90ed815031c2b00b259a7ca571/cmake/env/thisbdm.sh#L374" rel="nofollow noopener">https://github.com/BioDynaMo/biodynamo/blob/5092924f4caf7a90ed815031c2b00b259a7ca571/cmake/env/thisbdm.sh#L374</a></li>
<li>executing line <code> . scl_source enable llvm-toolset-6.0</code>
</li>
<li>failed line at the <code>scl_source</code> code</li>
</ol>
<pre><code class="lang-auto"># Now check if the collection isn't already enabled
/usr/bin/scl_enabled $arg &gt; /dev/null 2&gt; /dev/null      &lt;- failed line
    if [ $? -ne 0 ]; then
for     _scls+=($arg)
    _scl_prefixes+=($_scl_prefix)
    fi;
done
</code></pre>
<p>Fix:</p>
<pre><code class="lang-auto">run: |
        . scl_source enable devtoolset-7 || true
        . scl_source enable llvm-toolset-6.0 || true
        . /etc/profile.d/modules.sh || true
        module load mpi
        . $INSTALL_DIR/bin/thisbdm.sh
        git config --system user.name "Test User"
        git config --system user.email user@test.com
        xvfb-run -d -s "-screen 0 2560x1440x24" make run-demos
</code></pre>
<p><img src="https://user-images.githubusercontent.com/47745270/181094223-dc1ab442-508e-4c24-9ba6-5376ac8588b0.png" alt="image" data-base62-sha1="j34eDhemWADdYPQmZT9C28rIiFl" width="314" height="353"></p>

**Comment by Senui** [comment]
<p>Hi <a class="mention" href="/al-cheb">@al-cheb</a>,</p>
<p>Many thanks for investigating the issue at hand!</p>
<p>Although I believe that the line that you marked with “failed line” is by design supposed to return a non-zero exit code if <code>$arg</code> is not enabled. In which case, any script that uses non-zero return codes as a means to implement some logic will always fail a Github Action workflow…</p>
<p>Is this something you guys at Github considered? Will things stay like this?</p>
<p>Furthermore, it is strange that this was not reproducible on the same docker container on my own machine. Do you have any idea why this only fails on GH Actions?</p>

**Comment by al-cheb** [comment]
<p><a class="mention" href="/senui">@senui</a>,</p>
<p>Case with source:<br>
The  <code>/usr/bin/scl_enabled llvm-toolset-6.0</code>  encounters a  <code>exit 1</code>  which is different from zero. It will terminate the  <code>scl_source code</code>  instantaneously due to the  <code>exit</code>  statement and that’s why:<br>
<code>. $INSTALL_DIR/bin/thisbdm.sh</code> - failed and <code>$INSTALL_DIR/bin/thisbdm.sh</code> - works</p>
<p>Looks like it by design:</p>
<blockquote>
<p>return [n]</p>
<p>Causes a function to exit with the return value specified by n. If used outside a function, but during execution of a script by the . (source) command, it causes the shell to stop executing that script and return either n or the exit status of the last command executed within the script as the exit status of the script. If used outside a function and not during execution of a script by ., the return status is false.</p>
</blockquote>
<pre><code class="lang-auto">source filename [arguments]
    Read and execute commands from filename in the current shell environment and
    return **the exit status of the last command executed from filename**. 
</code></pre>
<aside class="onebox stackexchange">
  <header class="source">
      <a href="https://askubuntu.com/questions/182012/is-there-a-difference-between-and-source-in-bash-after-all" target="_blank" rel="nofollow noopener">askubuntu.com</a>
  </header>
  <article class="onebox-body">
      <a href="https://askubuntu.com/users/26972/ysap" target="_blank" rel="nofollow noopener">
    <img alt="ysap" src="https://user-images.githubusercontent.com/47745270/181094225-78ca84fe-e742-4a00-8536-f7ba6c3a71fb.jpeg" class="thumbnail onebox-avatar" width="128" height="128">
  </a>
<h4>
  <a href="https://askubuntu.com/questions/182012/is-there-a-difference-between-and-source-in-bash-after-all" target="_blank" rel="nofollow noopener">Is there a difference between "." and "source" in bash, after all?</a>
</h4>

<div class="tags">
  <strong>bash, scripts, environment-variables, environment</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://askubuntu.com/users/26972/ysap" target="_blank" rel="nofollow noopener">
    ysap
  </a>
  on <a href="https://askubuntu.com/questions/182012/is-there-a-difference-between-and-source-in-bash-after-all" target="_blank" rel="nofollow noopener">01:07AM - 30 Aug 12 UTC</a>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Reproduce behavior:</p>
<p>Without source( ./test.sh):</p>
<pre><code class="lang-auto">steps:
      - name: Test
        
        run: |
          cat &lt;&lt;EOF &gt; $HOME/subscript.sh
          #!/bin/sh
          exit 1
          EOF
          chmod +x $HOME/subscript.sh
          
          cat &lt;&lt;EOF &gt; test.sh
          #!/bin/bash
          $HOME/subscript.sh
          if [ $? -ne 0 ]; then
              echo PASS
          fi;
          EOF
          
          chmod +x ./test.sh
          echo START
          ./test.sh
          echo END
</code></pre>
<p><img src="https://user-images.githubusercontent.com/47745270/181094226-db99698a-53f0-4183-b97e-6d53ee03536e.png" alt="image" data-base62-sha1="zi58Hos0HmjKcFtvIzZHPEC9v80" width="352" height="131"></p>
<p>With source(source ./test.sh):</p>
<pre><code class="lang-auto">steps:
      - name: Test
        
        run: |
          cat &lt;&lt;EOF &gt; $HOME/subscript.sh
          #!/bin/sh
          exit 1
          EOF
          chmod +x $HOME/subscript.sh
          
          cat &lt;&lt;EOF &gt; test.sh
          #!/bin/bash
          $HOME/subscript.sh
          if [ $? -ne 0 ]; then
              echo PASS
          fi;
          EOF
          
          chmod +x ./test.sh
          echo START
          source ./test.sh
          echo END
</code></pre>
<p><img src="https://user-images.githubusercontent.com/47745270/181094227-2b8c031b-a01c-4178-afe3-115b5bd62f6c.png" alt="image" data-base62-sha1="mCt1E2X2pfESfiwIaZ1mlz7sPa9" width="400" height="166"></p>
<p>With source || true (source ./test.sh || true):</p>
<pre><code class="lang-auto">    steps:
      - name: Test
        
        run: |
          cat &lt;&lt;EOF &gt; $HOME/subscript.sh
          #!/bin/sh
          exit 1
          EOF
          chmod +x $HOME/subscript.sh
          
          cat &lt;&lt;EOF &gt; test.sh
          #!/bin/bash
          $HOME/subscript.sh
          if [ $? -ne 0 ]; then
              echo PASS
          fi;
          EOF
          
          chmod +x ./test.sh
          echo START
          source ./test.sh || true
          echo END
</code></pre>
<p><img src="https://user-images.githubusercontent.com/47745270/181094228-44f05721-1481-4484-bf1e-5c44950be8fc.png" alt="image" data-base62-sha1="uWTDoDbNdnQdpC3AQrQL7J0z75w" width="342" height="479"></p>

**Comment by Senui** [comment]
<p>Hi <a class="mention" href="/al-cheb">@al-cheb</a>,</p>
<p>Thanks for the clear explanation. I understand the situation better now, and will keep this in the back of my mind if I ever get a non-zero exit code again!</p>
<p>Thanks,<br>
Ahmad</p>
